### PR TITLE
JSON-API: 'included' and 'linkage'

### PR DESCRIPTION
**Pull request by danelowe**
Further changes to bring JSONAPI formatting more in line with 1.0 format

- Changed `linked` to `included`
- Change format of `links` to include 'linkages'

### CODE REVIEW

File: yaks/lib/yaks/format/json_api.rb [hunk 1]
```diff
@@ -38,7 +38,7 @@ def serialize_resource(resource)
       def serialize_links(subresources)
         subresources.each_with_object({}) do |resource, hsh|
           next if resource.null_resource?
-          hsh[resource.type] = serialize_link(resource)
+          hsh[resource.rels.first] = serialize_link(resource)
```
Comment (danelowe) [inline yaks/lib/yaks/format/json_api.rb]: @plexus I'm not sure if I've completely understood what rels is for. Should I perhaps be tagging rels e.g.

````ruby
['parent_association_name:preferred_works']
```

The idea being that the association name will not always match up with the resource type, so can't rely on resource.type.

I'll dig into the how the mappers work a bit more to make sure I've got this.

Comment (plexus) [inline yaks/lib/yaks/format/json_api.rb]: `rels` at the resource level means the relationship a resource has to its parent. Originally we stored subresources as a `{rel => resource}` hash, but some formats explicitly allow mutliple rels, so it was moved into the child resource, which I think makes for a better modeling in general.

Does that help you answer your question?

I'm mulling over some changes to make rels/identifiers in Yaks smarter in general, because different formats might represent things in different ways, e.g.

```
"http://api.example.com/rel/widgets"
"rel:widgets"
"widgets"
```

in that case a `rel` will be some object that knows how to represent itself as a full URI, a compacted CURIE form, just a "short name", etc. For a format like JSON-API where (I think) you would just use simple type/tag names instead of full URI's this would help. (It's also an important preparatory step to supporting RDF based stuff like JSON-LD).

**Comment by plexus** [comment]
And thank you for the pull request <3

File: yaks/lib/yaks/format/json_api.rb (hunk 1 above)
Comment (danelowe) [inline yaks/lib/yaks/format/json_api.rb]: So yeah I just read the manual...

> `:rel` : Set the relation (symbol or URI) this association has with the object. Will be derived from the association name and the configured rel_template if ommitted

In the context of JSON-API, I would assume that I would do something like this in the mapper:

```ruby
class CustomerMapper < Yaks::Mapper
# ...
  has_many :books, rel: :favourite_books
# ...
end
```

Loading a customer record and calling `yaks.map` on it, I looked at the subresources, and it had a `rels=[:favourite_books]` on the Yaks::CollectionResource. However if `rel:` is not defined in the mapper then the sub resource will have `rels=[]`, which would mean **I broke the formatter.** 

Any suggestions for where I could start adding more tests? Also should rels not be empty at this stage? I'm assuming that the formatter should not know or care about the mapper.

**Comment by danelowe** [comment]
My pleasure :) Yaks is extremely helpful.

File: yaks/lib/yaks/format/json_api.rb (hunk 1 above)
Comment (plexus) [inline yaks/lib/yaks/format/json_api.rb]: That sounds like a bug. You're right assuming it should have a (computed) rel if no explicit rel is provided.

`:favourite_books` would be an illegal rel, since according to RFC5899 they would have to be either a registered type (either in the [IANA](http://www.iana.org/assignments/link-relations/link-relations.xhtml) or [microformats](http://microformats.org/wiki/existing-rel-values) registry), or a fully qualified URI.

JSON-API doesn't confirm to RFC5899 though, so you would want to turn those URIs back into keywords, which is why the forementioned, but not yet implemented, "polymorphic" identifiers would come in handy.

Comment (plexus) [inline yaks/lib/yaks/format/json_api.rb]: I mean RFC5988, not 5899 :)

File: yaks/spec/json/confucius.json_api.json [hunk 2]
```diff
@@ -8,20 +8,31 @@
...
+    },
+    {
+      "id": 7,
+      "type": "scholars",
+      "href": "http://literature.example.com/authors/tolkein",
+      "name": "JRR",
+      "pinyin": "Joey",
+      "latinized": "Jorge",
+      "links": {
+        "works": {"linkage": [{"type": "works", "id": 17}]}
```
Comment (danelowe) [inline yaks/spec/json/confucius.json_api.json]: I feel kinda bad about this...

**Comment by plexus** [comment]
@danelowe could you provide a full example for the issue with `rel` you describe above? This is what I get

``` ruby
$y = Yaks.new {}

class Customer ; include Attribs.new(:books) ; end
class Book ; include Attribs.new(:title) ; end

class CustomerMapper < Yaks::Mapper
  has_many :books
end

class BookMapper < Yaks::Mapper
  attributes :title
end

puts $y.map(Customer.new(books: [Book.new(title: "Freedom")])).pp
```

output:

``` ruby
Yaks::Resource.new(
  type: "customer",
  subresources: [
    Yaks::CollectionResource.new(
      type: "book",
      rels: ["rel:books"],
      members: [
        Yaks::Resource.new(type: "book", attributes: {:title=>"Freedom"})
      ]
    )
  ]
)
```

**Comment by plexus** [comment]
I'm happy to merge this, but it will need a rebase. Do you want to do that or shall I do it?

**Comment by danelowe** [comment]
@plexus Apologies, it turns out the rels issue was simply that I had rel_template in my Yaks config. I originally used this to get started with Yaks: https://github.com/sinetris/yaks_grape_mongoid_example/blob/master/config/initializers/yaks.rb

I think I've done the rebase correctly.

**Comment by plexus** [comment]
> it turns out the rels issue was simply that I had rel_template in my Yaks config

That still sounds like a bug, having a `rel_template` in your settings should simply mean that it uses that template when no explicit rel is given on an association. I'll see if I can reproduce. A self contained example would still be very much appreciated.

I merged the PR. Sorry it's taken awhile, I'm a bit backed up on my open source stuff.

**Comment by danelowe** [comment]
@plexus Sorry for the slow reply, and even more sorry for pointing in the wrong direction. 

The config I had that was causing the behaviour was this line:

https://github.com/sinetris/yaks_grape_mongoid_example/blob/master/config/initializers/yaks.rb#L4
`rel_template 'rels:{dest}'`

I just searched the yaks repository and found this: 

https://github.com/plexus/yaks/blob/83da834d614ed837b0a41d5706b3d0f03fac9f4a/yaks/lib/yaks/breaking_changes.rb#L73
`When specifying a rel_template, now a single {rel} placeholder is
expected instead of {src} and {dest}.`

I also just added the line back into my config and ran the debugger. 

On a subresource, I actually got

`rels=["rels:"]`

Which is entirely consistent with what I would expect when using a placeholder in the template that no longer exists. 

I just checked the date on that example repository. More than 6 months ago. Oops.

**Comment by danelowe** [comment]
@plexus Also as per original comment, the template config I had was only causing issues where the rel wasn't explicitly given in the mapper.

**Comment by plexus** [comment]
Ok, that seems consistent with what I would expect :) thanks for looking it up!
